Statistics for precision of time data of local chrony

domschl opened the pull request:
Following the suggestions in #2357, here a module for chrony based on ExecutableService.

Please comment.

ktsaou commented:
@domschl thanks!
Waiting for @l2isbad to review this.

l2isbad commented on python.d/chrony.chart.py:
```diff
@@ -0,0 +1,92 @@
+        """
+        try:
+            chrony_dict = {ln.split(':', 1)[0].strip(): ln.split(':', 1)[1].strip().split(' ')[0]
+                           for ln in self._get_raw_data()[1:]}
```
`self._get_raw_data()` [returns](https://github.com/firehol/netdata/blob/master/python.d/python_modules/base.py#L919-L921) `None` if there is any problem.
```python
>>> None[1:]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object has no attribute '__getitem__'
```

Also dict comprehensions doesn't work in python 2.6

domschl commented:
@l2isbad thx for the input! Please have a look, if that addresses the points: I've added checks for None-return vals of _get_raw_data() and expanded the dict comprehension.

domschl commented:
Added some more sanity checks in case chronyc changes output format.

l2isbad commented:
The whole method in `try/except` block doesn't looks cool 😄 

I install chrony and the output from "chronyc -n tracking":

```
Reference ID    : 89.109.251.22 (89.109.251.22)
Stratum         : 2
Ref time (UTC)  : Mon Sep  4 08:57:56 2017
System time     : 0.000184203 seconds slow of NTP time
Last offset     : -0.000071091 seconds
RMS offset      : 0.001949425 seconds
Frequency       : 17.784 ppm fast
Residual freq   : -0.004 ppm
Skew            : 0.242 ppm
Root delay      : 0.111796 seconds
Root dispersion : 0.002530 seconds
Update interval : 1029.9 seconds
Leap status     : Normal
```
Or 

```
Reference ID    : 127.127.1.1 ()
Stratum         : 10
Ref time (UTC)  : Mon Sep  4 09:18:40 2017
System time     : 0.000000000 seconds fast of NTP time
Last offset     : +0.000000000 seconds
RMS offset      : 0.000000000 seconds
Frequency       : 17.784 ppm fast
Residual freq   : +0.000 ppm
Skew            : 0.000 ppm
Root delay      : 0.000000 seconds
Root dispersion : 0.000001 seconds
Update interval : 0.0 seconds
Leap status     : Not synchronised
```

Do we need to check "Leap status"? If not this is my example (not tested but should work)


```python
CHRONY = [('Frequency', 'frequency', 1e3),
          ('Last offset', 'lastoffset', 1e9),
          ('RMS offset', 'rmsoffset', 1e9),
          ('Residual freq', 'residualfreq', 1e3),
          ('Root delay', 'rootdelay', 1e9),
          ('Root dispersion', 'rootdispersion', 1e9),
          ('Skew', 'skew', 1e3),
          ('System time', 'timediff', 1e9)]
....

def _get_data(self):
    raw_data = self._get_raw_data()
    if not raw_data:
        return None
    
    raw_data = (line.split(':', 1) for line in raw_data)
    parsed, data = dict(), dict()

    for line in raw_data:
        try:
            key, value = (x.strip() for x in line)
        except ValueError:
            continue
        parsed[key] = value.split()[0]

    for key, dim_id, multiplier in CHRONY:
        try:
            data[dim_id] = int(float(raw[key]) * multiplier)
        except KeyError:
            continue

    return data or None

```

l2isbad commented on python.d/chrony.chart.py:
```diff
@@ -0,0 +1,108 @@
+              ('Root delay', 'rootdelay', 1e9),
+              ('Root dispersion', 'rootdispersion', 1e9),
+              ('Skew', 'skew', 1e3),
+              ('System time', 'timediff', 1e9)]
```
Move the `CHRONY` list outside the Service class.

domschl replied:
will do...

l2isbad commented on python.d/chrony.chart.py:
```diff
@@ -0,0 +1,108 @@
+                key, value = (l.strip() for l in line)
+            except ValueError:
+                continue
+            if len(value) > 0:
```
In what cases do we need this check?

domschl replied:
you are too fast ;-) S.b.

l2isbad replied:
> value.split()[0] crashes on value=''

then
```python
if value:
    ...
```
?
No reason to len() (if i understand it right :smile:)

domschl replied:
right you are! Done.

domschl commented:
@l2isbad thanks again! I've implemented your suggestions with three small changes:
* typo: `int(float(raw[key]) * multiplier)`   needs to be `int(float(parsed[key]`...
* `value.split()[0]` crashes on `value=''` with index out-of-range, fixed with check, in case chrony sends a line ending with ':' (e.g. 'Frequency :').
* `float(parsed[key])` crashes, if `parsed[key]` is non-numeric, added exception-type (e.g. if chrony sends 'Frequency : not initialized').

l2isbad commented:
looks ok @ktsaou 
thx @domschl 
And if you have time please do:
1. Add chrony to https://github.com/firehol/netdata/wiki/Add-more-charts-to-netdata
2. https://github.com/firehol/netdata/blob/master/python.d/README.md

l2isbad commented:
And
3. https://github.com/firehol/netdata/blob/master/conf.d/python.d.conf

And you have to update Makefiles or it will not be installed:
1. https://github.com/firehol/netdata/blob/master/conf.d/Makefile.am
2. https://github.com/firehol/netdata/blob/master/python.d/Makefile.am

domschl commented:
Concerning 'leap status': this just gives textual information about possible upcoming leap seconds. This can only happen twice a year, and isn't something that can be plotted in a useful way. Possible values are: `Normal`, `Insert second`, `Delete second` or `Not synchronized`. So for now, I don't do anything with those.

Makefiles & wik, doc: I will do the updates.

domschl commented:
Updates to Makefile, conf and documentation done. Please let me know if there's anything else to be done.

@l2isbad : thanks a lot for the help and guidance.

l2isbad commented:
@ktsaou ready

ktsaou commented:
merged! thanks!
